### PR TITLE
perf: Parallelize attachToTangle API for improving performance

### PR DESCRIPTION
**Pull request by marktwtn**
## Description

DO NOT MERGE

As the dcurl introduces remote interface of PoW calculation,
the attachToTangle API should be parallelized and distribute the PoWs
to the multiple remote hardwares such as FPGA boards.

## Type of change

- Enhancement (a non-breaking change which adds functionality)

### CODE REVIEW

File: src/main/java/com/iota/iri/service/API.java [hunk 1]
```diff
@@ -1489,15 +1492,18 @@ public static void incEllapsedTimePoW(long ellapsedTime) {
...
-                API.incEllapsedTimePoW(System.nanoTime() - startTime);
-                API.incCounterPoW();
-                if ( ( API.getCounterPoW() % 100) == 0 ) {
-                    String sb = "Last 100 PoW consumed "
-                                + API.getEllapsedTimePoW() / 1000000000L
-                                + " seconds processing time.";
-                    log.info(sb);
-                    counter_PoW = 0;
-                    ellapsedTime_PoW = 0L;
+                endTime = System.nanoTime();
```
Comment (jserv) [inline src/main/java/com/iota/iri/service/API.java]: Move the scope of `endTime` here. That is, declare it within `finally` block.

**Comment by jserv** [comment]
How can you evaluate the preliminary benefit with this change?

File: src/main/java/com/iota/iri/service/API.java [hunk 2]
```diff
@@ -120,6 +120,8 @@
 
     private final String[] features;
 
+    private static Object lock = new Object();
```
Comment (jserv) [inline src/main/java/com/iota/iri/service/API.java]: I am afraid the naming "lock" is misleading.

**Comment by marktwtn** [comment]
> How can you evaluate the preliminary benefit with this change?

I will evaluate the benefit in https://github.com/DLTcollab/general/issues/162.

File: src/main/java/com/iota/iri/service/API.java (hunk 2 above)
Comment (marktwtn) [reply src/main/java/com/iota/iri/service/API.java]: I was thinking that it should be acting like a lock.
I want to know about you concern more that I can come up with a better naming.

Comment (jserv) [reply src/main/java/com/iota/iri/service/API.java]: There might be more than 2 locks within this class.

File: src/main/java/com/iota/iri/service/API.java (hunk 1 above)
Comment (marktwtn) [reply src/main/java/com/iota/iri/service/API.java]: Done.

File: src/main/java/com/iota/iri/service/API.java (hunk 2 above)
Comment (marktwtn) [reply src/main/java/com/iota/iri/service/API.java]: I change it from `lock` to `PoWTimeLock`.

**Review by wusyong — APPROVED** [review]

**Pull request merged**